### ISSUE TITLE
An error raised in an ActiveDelivery line behaves differently than one raised in an ActiveMailer line

### ISSUE THREAD

**Issue by thornomad** [palkan/abstract_notifier#9]
Am using `ActiveDelivery` in combination with `AbstractNotifier`.  Thanks so much for both contributions.  I have a few custom abstract notifier lines setup as well as a mailer line (default).  Great results!

I have a question/issue with error handling and job queueing.

What I am observing is that when an error is raised inside the `AbstractNotifier` it halts the execution of all the other delivery lines during the process of the delivery (even when using `#notify` to _enqueue_ the jobs).  Whereas if an error is raised inside of an `ActiveMailer` line the error does _not_ stop the delivery lines from being enqueued (and instead the error actually gets raised when the enqueued job runs).

Simple example:

```ruby
class FooBarDelivery < ApplicationDelivery; end

class FooBarNotifier < AbstractNotifier::Base
  def foo
    raise "This will halt the operating of `FooBarDelivery.notify(:foo)`"
  end
end

class FooBarMailer < ApplicationMailer
  def foo
    raise "This will *not* halt the operation of `FooBarDelviery.notify(foo)`"
  end
end

# => FooBarDelivery.with(foo: :bar).notify(:foo)
```

So the raised error in the abstract notifier halts everything; the raised error in ApplicationMailer does not.  This is tripping me up in particular with nested job queues because I had one abstract notifier line that would error but all the other lines were being _retried_ resulting in a lot of "spam" (as the one line never was getting delivered).  

I would have expected them all to behave like ActiveMailer — they aren't actually executed until after the job is queued.  

I wonder if I have my driver setup wrong or is this just how it happens to be?

Thanks!

**Comment by palkan** [comment]
Hey @thornomad!

Your setup looks fine.

> I would have expected them all to behave like ActiveMailer — they aren't actually executed until after the job is queued.

There is a difference in the way ActionMailer and AbstractNotifier. Let's see how they work:
- ActionMailer creates an intermediate [MessageDelivery](https://github.com/rails/rails/blob/main/actionmailer/lib/action_mailer/message_delivery.rb) object whenever you call `MyMailer.some_method`; then you call `#deviler_later`, which enqueues the job with this message delivery object; the action itself hasn't been processed yet, we pass class, params and arguments to the job and only invoke the action inside it.
- AbstractNotifier also uses an intermediate object, [Notification](https://github.com/palkan/abstract_notifier/blob/103ce1f11a08f86e30f43849ee83f2ba6d186bf5/lib/abstract_notifier/base.rb#L7), but unlike ActionMailer, this object performs the action immediately to calculate the payload for the notifier adapter. Thus, we only enqueue the notifier class (to get the adapter) and the final payload.

tl;dr ActionMailer lazily calculate the payload, while AbstractNotifier does not.

That was intentional at the time I've extracted this library from the commercial project: notifier payloads were pretty small and cheap to calculate, so it didn't make any sense to calculate them lazily (and thus, make the implementation a bit more complex).

Can we refactor the internals to process notifications lazily? Yeah, I think, we can. Probably, we should make it configurable per notifier class.

**Comment by thornomad** [comment]
Hi @palkan — thanks for the detailed response.  Your description fits with the behavior I was seeing about "immediately calculating the payload".  For our setup, we have an ActiveJob is calling `FooBarDelivery.notify` _at the end of it_.  If there is an error during the _immediate_ part of the execution in an `AbstractDelivery` line it is causing the whole job to re-run (and all the lines re-run up to the failure) which is resulting in a lot of spam notifications.  `ActiveMailer`, as you point out, doesn't have this behavior.

For the project I am working on, we are going to need the lazy execution (one way or another).  I haven't dug too much into the codebase yet, but if you are willing to have a flag that also you to designate some notifiers as "lazy" versus others I'm happy to look into sorting it out.    And if you have an idea about how it could be done effectively let me know!

**Comment by dmitry** [comment]
@thornomad @palkan would be great to continue to work on this issue. Do you have an existing solutions, which can be embedded into this gem?

**Comment by thornomad** [comment]
@dmitry — I ended up stopping using `ActiveDelivery` and just rolled my own delivery line based on the recommendations at the `ActiveDelivery` README.  I didn't see a clear pathway to easily converting this gem to do what I wanted without more work than just writing our own.  Probably need to break this into two different classes/modules that a user can inherit from like `AbstractNotifier::Base` and `AbstractNotifer::LazyBase` or something.  Heh.

**Comment by palkan** [comment]
> Probably need to break this into two different classes/modules that a user can inherit from like AbstractNotifier::Base and AbstractNotifer::LazyBase or something

I like it. Though, maybe, we can do a little bit different thing: build a new _lazy line_  and use it for AbstractNotifier.
Smth like this:

```ruby
# stays the same
def notify_now(handler, mid, *args)
  handler.public_send(mid, *args).notify_now
end

# use a separate job
def notify_later(handler, mid, *args)
   handler.class.async_adapter.enqueue_raw(handler.class, params, notification_name, args)
end

module AbstractNotifier
  module AsyncAdapters
    class ActiveJob
      class RawDeliveryJob < ::ActiveJob::Base
        def perform(notifier_class_name, params, action, payload)
           notifier_class = notifier_class_name.constantize
           notifier = notifier_class.new(action, **params)
           notifier.public_send(action, *payload).notify_now
        end
      end
    end
  end
end
```
